locator/abstract_replication_strategy.hh

```diff
@@ -128,11 +128,48 @@ public:
     future<dht::token_range_vector> get_pending_address_ranges(const token_metadata_ptr tmptr, std::unordered_set<token> pending_tokens, inet_address pending_address, locator::endpoint_dc_rack dr) const;
 };
 
+using replication_strategy_ptr = seastar::shared_ptr<const abstract_replication_strategy>;
+using mutable_replication_strategy_ptr = seastar::shared_ptr<abstract_replication_strategy>;
+
+/// \brief Represents effective replication (assignment of replicas to keys).
+///
+/// It's a result of application of a given replication strategy
+/// over a particular token metadata version, for a given table.
+/// Can be shared by multiple tables if they have the same replication.
+///
+/// Immutable, users can assume that it doesn't change.
+///
+/// Holding to this object keeps the associated token_metadata_ptr alive,
+/// keeping the token metadata version alive and seen as in use.
+class effective_replication_map {
+protected:
+    replication_strategy_ptr _rs;
+    token_metadata_ptr _tmptr;
+    size_t _replication_factor;
+public:
+    effective_replication_map(replication_strategy_ptr, token_metadata_ptr, size_t replication_factor);
```

Comment:
`noexcept` right? (token_effective_replication_map ctor is noexcept today)

Reply:
Not any more. storage_service::replicate_to_all_cores handles exceptions when creating / copying token_metadata and effective_replication_map.
We're in disagreement on this.  I think that noexcept is beneficial when possible, not only when required.

Reply:
We shouldn't block the PR on this, but since `effective_replication_map` was already marked noexcept I suggest not to change that "on the fly" without verifying that it doesn't expose any caller that relies on it.
That's the point behind noexcept.  It shouldn't be changed on a whim.

Reply:
> It shouldn't be changed on a whim.

True, I'll preserve the noexcept.